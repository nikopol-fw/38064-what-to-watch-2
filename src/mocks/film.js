export const film = {
  title: `The Grand Budapest Hotel`,
  genre: `Drama`,
  releaseYear: `2014`,
  posterImg: `bg-the-grand-budapest-hotel.jpg`,
  coverImg: `the-grand-budapest-hotel-poster.jpg`,
};
