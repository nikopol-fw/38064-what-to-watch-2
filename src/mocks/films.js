export const films = [{
  title: `Fantastic Beasts: The Crimes of Grindelwald`,
  img: `fantastic-beasts-the-crimes-of-grindelwald.jpg`,
}, {
  title: `Bohemian Rhapsody`,
  img: `bohemian-rhapsody.jpg`,
}, {
  title: `Macbeth`,
  img: `macbeth.jpg`,
}, {
  title: `Aviator`,
  img: `aviator.jpg`,
}, {
  title: `We need to talk about Kevin`,
  img: `we-need-to-talk-about-kevin.jpg`,
}];
