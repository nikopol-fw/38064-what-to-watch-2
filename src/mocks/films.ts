import {Film} from "../models/Film";


export const films: Film[] = [{
  backgroundColor: `#A6B7AC`,
  backgroundImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/background/gangs_of_new_york.jpg`,
  description: `In 1862, Amsterdam Vallon returns to the Five Points area of New York City seeking revenge against Bill the Butcher, his father's killer.`,
  director: `Martin Scorsese`,
  genre: `Crime`,
  id: 1,
  isFavorite: false,
  name: `Gangs of new york`,
  posterImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/Gangs_of_New_York_Poster.jpg`,
  previewImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/gangs_of_new_york.jpg`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  rating: 8.8,
  released: 2002,
  runTime: 167,
  scoresCount: 370881,
  starring: [`Leonardo DiCaprio`, `Cameron Diaz`, `Daniel Day-Lewis`],
  videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
}, {
  backgroundColor: `#9B7E61`,
  backgroundImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg`,
  description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
  director: `Steven Spielberg`,
  genre: `Adventure`,
  id: 2,
  isFavorite: false,
  name: `War of the Worlds`,
  posterImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg`,
  previewImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  rating: 9.3,
  released: 2005,
  runTime: 116,
  scoresCount: 386834,
  starring: [`Tom Cruise`, `Dakota Fanning`, `Tim Robbins`],
  videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
}, {
  backgroundColor: `#F1E9CE`,
  backgroundImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/background/Macbeth.jpg`,
  description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
  director: `Justin Kurzel`,
  genre: `Drama`,
  id: 3,
  isFavorite: false,
  name: `Macbeth`,
  posterImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/Macbeth.jpg`,
  previewImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/macbeth.jpg`,
  previewVideoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  rating: 3.3,
  released: 2015,
  runTime: 113,
  scoresCount: 48798,
  starring: [`Michael Fassbender`, `Marion Cotillard`, `Jack Madigan`],
  videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
}, {
  backgroundColor: `#92918B`,
  backgroundImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/background/Revenant.jpg`,
  description: `A frontiersman on a fur trading expedition in the 1820s fights for survival after being mauled by a bear and left for dead by members of his own hunting team.`,
  director: `Alejandro G. Iñárritu`,
  genre: `Action`,
  id: 4,
  isFavorite: false,
  name: `The Revenant`,
  posterImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/Revenant.jpg`,
  previewImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/revenant.jpg`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  rating: 4,
  released: 2015,
  runTime: 156,
  scoresCount: 618498,
  starring: [`Leonardo DiCaprio`, `Tom Hardy`, `Will Poulter`],
  videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
}, {
  backgroundColor: `#F0DBA2`,
  backgroundImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/background/Johnny_English.jpg`,
  description: `After a sudden attack on the MI5, Johnny English, Britain's most confident yet unintelligent spy, becomes Britain's only spy.`,
  director: `Peter Howitt`,
  genre: `Comedy`,
  id: 5,
  isFavorite: false,
  name: `Johnny English`,
  posterImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/Johnny_English.jpg`,
  previewImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/johnny-english.jpg`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  rating: 10,
  released: 2003,
  runTime: 88,
  scoresCount: 136843,
  starring: [`Rowan Atkinson`, `John Malkovich`, `Natalie Imbruglia`],
  videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
}];
