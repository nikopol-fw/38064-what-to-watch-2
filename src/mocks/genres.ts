export const genres = [`All genres`, `Action`, `Adventure`, `Comedy`, `Crime`, `Drama`, `Fantasy`, `Thriller`];
