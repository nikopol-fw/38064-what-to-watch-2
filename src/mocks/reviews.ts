import {Review} from "../models/review";


export const reviews: Review[] = [{
  id: 1,
  rating: 1.3,
  comment: `Poised to be an instant classic, almost everything about this film is phenomenal - the acting, the cinematography, the discography, etc.`,
  date: new Date(`2019-12-05T19:57:19.384Z`),
  user: {
    id: 16,
    name: `Mollie`,
  }
}, {
  id: 2,
  rating: 2.7,
  comment: `This movie really touched my heart, it really is the best movie of the year and everyone should see this masterpiece.`,
  date: new Date(`2019-11-30T19:57:19.384Z`),
  user: {
    id: 12,
    name: `Isaac`,
  }
}];
