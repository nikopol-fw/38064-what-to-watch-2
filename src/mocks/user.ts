import {User} from "../models/user";


export const user: User = {
  avatar: `/wtw/static/avatar/3.jpg`,
  email: `alter.ego@gmail.com`,
  id: 1,
  name: `alter.ego`,
};
