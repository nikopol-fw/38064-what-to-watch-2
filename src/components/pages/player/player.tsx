import * as React from 'react';
import {RefObject} from 'react';
import {compose} from 'redux';
import {connect} from 'react-redux';
import {RouteComponentProps} from 'react-router-dom';

import {Film} from '../../../models/film';
import {getFilmById} from '../../../reducer/data/selectors';
import {withPlayerFunctionality} from '../../../hocs/with-player-functionality/with-player-functionality';
import {VideoPlayer} from '../../shared/video-player/video-player';
import {PlayerProgress} from '../../shared/player-progress/player-progress';


interface Props extends RouteComponentProps {
  film: Film;

  isPlaying: boolean;
  percentage: number;
  timingString: string;
  videoRef: RefObject<HTMLVideoElement>;
  onFullScreenBtnClick: () => void;
  onPlayBtnClick: () => void;
  onPauseBtnClick: () => void;
  onMetadataLoaded: () => void;
  onTimeUpdate: () => void;
}

export class Player extends React.PureComponent<Props> {

  static readonly defaultProps = {
    film: {
      backgroundColor: null,
      backgroundImage: null,
      description: null,
      director: null,
      genre: null,
      id: null,
      isFavorite: null,
      name: null,
      posterImage: null,
      previewImage: null,
      previewVideoLink: null,
      rating: null,
      released: null,
      runTime: null,
      scoresCount: null,
      starring: null,
      videoLink: null,
    },
    isPlaying: false,
    percentage: 0,
    timingString: ``,
  };

  constructor(props: Props) {
    super(props);

    this.handleExitBtnClick = this.handleExitBtnClick.bind(this);
  }

  private handleExitBtnClick() {
    const {history, location} = this.props;
    if (location.key) {
      history.goBack();
    } else {
      history.push(`/`);
    }
  }

  render() {
    const {
      film,
      isPlaying,
      percentage,
      timingString,
      videoRef,
      onFullScreenBtnClick,
      onPlayBtnClick,
      onPauseBtnClick,
      onMetadataLoaded,
      onTimeUpdate,
    } = this.props;

    return (
      <div className="player">
        <VideoPlayer
          isMuted={false}
          poster={film.previewImage}
          video={film.videoLink}
          videoRef={videoRef}
          onMetadataLoaded={onMetadataLoaded}
          onTimeUpdate={onTimeUpdate}
        />

        <button type="button" className="player__exit" onClick={this.handleExitBtnClick}>Exit</button>

        <div className="player__controls">

          <PlayerProgress
            percentage={percentage}
            timingString={timingString}
          />

          <div className="player__controls-row">
            {isPlaying
              ? (
                <button type="button" className="player__play" onClick={onPauseBtnClick}>
                  <svg viewBox="0 0 14 21" width="14" height="21">
                    <use xlinkHref="#pause"/>
                  </svg>
                  <span>Pause</span>
                </button>
              ) : (
                <button type="button" className="player__play" onClick={onPlayBtnClick}>
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s"/>
                  </svg>
                  <span>Play</span>
                </button>
              )
            }
            <div className="player__name">{film.name}</div>

            <button type="button" className="player__full-screen" onClick={onFullScreenBtnClick}>
              <svg viewBox="0 0 27 27" width="27" height="27">
                <use xlinkHref="#full-screen"/>
              </svg>
              <span>Full screen</span>
            </button>
          </div>
        </div>
      </div>
    );
  }
}


const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  film: getFilmById(state, +ownProps.match.params.id),
});


export default compose(
    connect(mapStateToProps),
    withPlayerFunctionality
)(Player);
