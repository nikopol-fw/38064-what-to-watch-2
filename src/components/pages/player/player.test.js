import React from 'react';
import renderer from 'react-test-renderer';
import {StaticRouter} from 'react-router-dom';

import {films} from '../../../mocks/films';
import {Player} from './player';


const mock = {
  film: films[0],
  isPlaying: false,
  percentage: 50,
  timingString: `00:07:47`,
  videoRef: React.createRef(),
  onFullScreenBtnClick: () => void (0),
  onPlayBtnClick: () => void (0),
  onPauseBtnClick: () => void (0),
  onMetadataLoaded: () => void (0),
  onTimeUpdate: () => void (0),
};

it(`Player correctly renders afters relaunch`, () => {
  const tree = renderer
    .create(
        <StaticRouter>
          <Player
            film={mock.film}
            isPlaying={mock.isPlaying}
            percentage={mock.percentage}
            timingString={mock.timingString}
            videoRef={mock.videoRef}
            onFullScreenBtnClick={mock.onFullScreenBtnClick}
            onPlayBtnClick={mock.onPlayBtnClick}
            onPauseBtnClick={mock.onPauseBtnClick}
            onMetadataLoaded={mock.onMetadataLoaded}
            onTimeUpdate={mock.onTimeUpdate}
          />
        </StaticRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
