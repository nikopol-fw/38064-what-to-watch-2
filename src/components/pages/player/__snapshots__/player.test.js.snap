// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player correctly renders afters relaunch 1`] = `
<div
  className="player"
>
  <video
    className="player__video"
    muted={false}
    onLoadedMetadata={[Function]}
    onTimeUpdate={[Function]}
    poster="https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/gangs_of_new_york.jpg"
    preload="metadata"
    src="http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4"
  />
  <button
    className="player__exit"
    onClick={[Function]}
    type="button"
  >
    Exit
  </button>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max="100"
          value={50}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "50%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        00:07:47
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlinkHref="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      <div
        className="player__name"
      >
        Gangs of new york
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;
